common/src/math.rs

```diff
@@ -53,79 +126,268 @@ impl<N: Scalar> Deref for MIsometry<N> {
     pub fn to_f32(self) -> MIsometry<f32> {
         MIsometry(self.0.cast::<f32>())
     }
 }
 
 impl MVector<f32> {
+    /// Casts the components to an `f64`
     pub fn to_f64(self) -> MVector<f64> {
         MVector(self.0.cast::<f64>())
     }
 }
 
 impl MVector<f64> {
+    /// Casts the components to an `f32`
     pub fn to_f32(self) -> MVector<f32> {
         MVector(self.0.cast::<f32>())
     }
 }
 
-impl<N: RealField + Copy> MIsometry<N> {
+impl MUnitPointVector<f32> {
+    /// Casts the components to an `f64`
     #[inline]
-    pub fn as_ref(self) -> [[N; 4]; 4] {
-        *self.0.as_ref()
+    pub fn to_f64(self) -> MUnitPointVector<f64> {
+        MUnitPointVector(self.0.to_f64())
     }
+}
+
+impl MUnitPointVector<f64> {
+    /// Casts the components to an `f32`
     #[inline]
-    pub fn unit_quaternion_to_homogeneous(rotation: na::UnitQuaternion<N>) -> Self {
-        MIsometry(rotation.to_homogeneous())
+    pub fn to_f32(self) -> MUnitPointVector<f32> {
+        MUnitPointVector(self.0.to_f32())
     }
+}
+
+impl MUnitDirectionVector<f32> {
+    /// Casts the components to an `f64`
     #[inline]
-    pub fn rotation_to_homogeneous(rotation: na::Rotation3<N>) -> Self {
-        MIsometry(rotation.to_homogeneous())
+    pub fn to_f64(self) -> MUnitDirectionVector<f64> {
+        MUnitDirectionVector(self.0.to_f64())
     }
+}
+
+impl MUnitDirectionVector<f64> {
+    /// Casts the components to an `f32`
     #[inline]
-    pub fn row(self, i: usize) -> na::RowVector4<N> {
-        self.0.row(i).into()
+    pub fn to_f32(self) -> MUnitDirectionVector<f32> {
+        MUnitDirectionVector(self.0.to_f32())
     }
+}
+
+impl<N: RealField + Copy> AsRef<[[N; 4]; 4]> for MIsometry<N> {
+    #[inline]
+    fn as_ref(&self) -> &[[N; 4]; 4] {
+        self.0.as_ref()
+    }
+}
+
+impl<N: RealField + Copy> From<na::UnitQuaternion<N>> for MIsometry<N> {
+    /// Converts a quaternion into the matrix for the rotation it represents.
+    fn from(value: na::UnitQuaternion<N>) -> Self {
+        MIsometry(value.to_homogeneous())
+    }
+}
+
+impl<N: RealField + Copy> From<na::Rotation3<N>> for MIsometry<N> {
+    /// Converts a rotation into the matrix representing that rotation.
+    fn from(value: na::Rotation3<N>) -> Self {
+        MIsometry(value.to_homogeneous())
+    }
+}
+
+impl<N: RealField + Copy> MIsometry<N> {
+    /// Returns a view containing the i-th row of this matrix.
+    #[inline]
+    pub fn row(&self, i: usize) -> na::MatrixView1x4<'_, N, na::U1, na::U4> {
+        self.0.row(i)
+    }
+
+    /// Creates an identity matrix.
     #[inline]
     pub fn identity() -> Self {
         Self(na::Matrix4::identity())
     }
-    #[inline]
-    pub fn map<N2: Scalar, F: FnMut(N) -> N2>(&self, f: F) -> MIsometry<N2> {
-        MIsometry(self.0.map(f))
+
+    /// The reflection about the hyperbolic plane represented by the given
+    /// normalized vector. An incorrect matrix will be returned if `a` and `b`
+    /// are not normalized.
```

Comment:
This comment is very confusing. There is no `a` or `b`, and if `normal` needs to be normalized, why isn't it one of the unit types?

Reply:
Oh, yeah, it's confusing because it's simply wrong. Sorry about that! I left a note in the original PR as well: https://github.com/Ralith/hypermine/pull/442/files#r1957455181